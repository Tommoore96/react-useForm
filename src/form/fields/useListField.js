import { List, Map } from 'immutable'
import { actions } from '../reducer/fieldReducer'
import { resolveFieldData, getFieldProps } from '../resolveFieldData'
import { getFields } from '../reducer/generateDefaultListState'
import { prepareNameForValidate } from './prepareNameForValidate'
/**
 * @param  fieldArgs { import("../useForm").IFormFieldArgs}
 * @returns {import("../useForm").IFormField}
 */
export const useListField = (state, dispatch, fieldArgs = {}, getAllValues) => {
  const requiredMessage = fieldArgs.requiredMessage || 'Required'

  const setValue = () => {
    // dispatch(actions.updateValue(fieldArgs.name, v))
    // tryValidate(v, state.getIn(['current', 'touched']))
  }

  const validate = (items) => {
    let message
    if (fieldArgs.validate) {
      message = fieldArgs.validate(items, prepareNameForValidate(fieldArgs.name), state.get(('getAllValues')))
    }
    if (!message && items.size === 0 && !fieldArgs.optional) {
      message = requiredMessage
    }

    dispatch(actions.validationResult(fieldArgs.name, Boolean(message), message || ''))
  }

  const add = (item = Map()) => {
    const options = state.getIn(['initial', 'options'])
    const field = state.getIn(['initial', 'field'])
    const items = state.get('items', List())
    const itemState = getFields(field.fields, item, options, fieldArgs.name, items.size)
    dispatch(actions.addListItem(fieldArgs.name, itemState))
    validate(items.push(item))
  }

  const remove = index => {
    const items = state.get('items', List())
    dispatch(actions.removeListItem(fieldArgs.name, index))
    validate(items.remove(index))
  }

  const updateIndex = (fromIndex, toIndex) => {
    dispatch(actions.updateListIndex(fieldArgs.name, fromIndex, toIndex))
  }

  const fieldData = state.getIn(['items'], List()).map(item => {
    return resolveFieldData(item, dispatch, getAllValues)
  }).toArray()

  return {
    props: {
      error: state.getIn(['current', 'error']),
      helperText: state.getIn(['current', 'helperText']),
      label: state.getIn(['initial', 'label']),
      items: fieldData.map((item, index) => ({
        ...getFieldProps(item, state.getIn(['items', index])),
        key: `${fieldArgs.name}.${index}`,
      })),
      add,
      remove,
      updateIndex,
    },
    setValue,
  }
}

export const defaultListValue = List()
