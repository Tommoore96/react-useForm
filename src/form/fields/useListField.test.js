import { fromJS } from 'immutable'
import { useListField } from './useListField'

const dispatch = jest.fn()
let state, fieldArgs
beforeEach(() => {
  fieldArgs = {
    name: 'fieldName',
    type: 'list',
    fields: [
      { name: 'first', label: 'First' },
    ],
  }

  state = fromJS({
    initial: {
      label: 'the label',
      type: 'list',
    },
    current: {
      error: false,
      helperText: 'help me',
    },
    items: [
      {
        fields: {
          first: {
            initial: {
              type: 'text',
              label: 'First',
              field: 'initialFirstField',
            },
            current: {
              error: 'false',
              helperText: '',
              value: '1',
            },
          },
          second: {
            initial: {
              type: 'text',
              label: 'Second',
              field: 'initialSecondField',
            },
            current: {
              error: 'false',
              helperText: '',
              value: '2',
            },
          },
        },
      },
    ],
  }).setIn(['initial', 'options'], {})
    .setIn(['initial', 'field'], fieldArgs)

})

test('returns props needed', () => {
  expect(useListField(state, dispatch, fieldArgs)).toMatchSnapshot()
})

test('add', () => {
  const { props } = useListField(state, dispatch, fieldArgs)
  props.add()
  expect(dispatch.mock.calls[0]).toMatchSnapshot()
})

test('add with values', () => {
  const { props } = useListField(state, dispatch, fieldArgs)
  props.add(fromJS({ first: '111' }))
  expect(dispatch.mock.calls[0]).toMatchSnapshot()
})

test('add will trigger validation', () => {
  const validate = jest.fn()
  validate.mockReturnValue('i am error')
  state = state.set('getAllValues', 'getAllValuesFunc')

  const { props } = useListField(state, dispatch, { ...fieldArgs, validate, name: 'colors.items.1.fields.shades.items.0.fields.shade' })
  props.add()
  expect(dispatch.mock.calls[1]).toMatchSnapshot()
  expect(validate.mock.calls[0]).toMatchSnapshot()
})

test('add will trigger custom validation', () => {
  const custom = jest.fn()
  custom.mockReturnValue('i am error')

  const { props } = useListField(state, dispatch, { ...fieldArgs, validate: custom })
  props.add()
  expect(dispatch.mock.calls[1]).toMatchSnapshot()
})

test('can handle list arrays that have no data', () => {
  const testState = state.deleteIn(['items'])
  expect(() => useListField(testState, dispatch, fieldArgs)).not.toThrowError()
})

test('remove', () => {
  const { props } = useListField(state, dispatch, fieldArgs)
  props.remove(1)
  expect(dispatch.mock.calls[0]).toMatchSnapshot()
})

test('remove will trigger validation', () => {
  const { props } = useListField(state, dispatch, fieldArgs)
  props.remove(1)
  expect(dispatch.mock.calls[1]).toMatchSnapshot()
})

test('remove will trigger custom validation', () => {
  const custom = jest.fn()
  custom.mockReturnValue('i am error')

  const { props } = useListField(state, dispatch, { ...fieldArgs, validate: custom })
  props.remove(1)
  expect(dispatch.mock.calls[1]).toMatchSnapshot()
})

it('updateIndex will dispatch action', () => {
  const { props } = useListField(state, dispatch, fieldArgs)
  props.updateIndex(0, 4)
  expect(dispatch.mock.calls[0]).toMatchSnapshot()
})
