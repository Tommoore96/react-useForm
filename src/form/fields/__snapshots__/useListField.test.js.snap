// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Immutable.Map {
      "fields": Immutable.Map {
        "first": Immutable.Map {
          "initial": Immutable.Map {
            "type": "text",
            "optional": false,
            "label": "First",
            "field": Object {
              "label": "First",
              "name": "fieldName.items.1.fields.first",
            },
            "value": "",
          },
          "current": Immutable.Map {
            "helperText": "",
            "error": false,
            "pristine": true,
            "touched": false,
            "value": "",
          },
        },
      },
    },
    "type": "addListItem",
  },
]
`;

exports[`add will trigger custom validation 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Object {
      "error": true,
      "helperText": "i am error",
    },
    "type": "validationResult",
  },
]
`;

exports[`add will trigger validation 1`] = `
Array [
  Object {
    "fieldName": "colors.items.1.fields.shades.items.0.fields.shade",
    "payload": Object {
      "error": true,
      "helperText": "i am error",
    },
    "type": "validationResult",
  },
]
`;

exports[`add will trigger validation 2`] = `
Array [
  Immutable.List [
    Immutable.Map {
      "fields": Immutable.Map {
        "first": Immutable.Map {
          "initial": Immutable.Map {
            "type": "text",
            "label": "First",
            "field": "initialFirstField",
          },
          "current": Immutable.Map {
            "error": "false",
            "helperText": "",
            "value": "1",
          },
        },
        "second": Immutable.Map {
          "initial": Immutable.Map {
            "type": "text",
            "label": "Second",
            "field": "initialSecondField",
          },
          "current": Immutable.Map {
            "error": "false",
            "helperText": "",
            "value": "2",
          },
        },
      },
    },
    Immutable.Map {},
  ],
  "colors.1.shades.0.shade",
  "getAllValuesFunc",
]
`;

exports[`add with values 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Immutable.Map {
      "fields": Immutable.Map {
        "first": Immutable.Map {
          "initial": Immutable.Map {
            "type": "text",
            "optional": false,
            "label": "First",
            "field": Object {
              "label": "First",
              "name": "fieldName.items.1.fields.first",
            },
            "value": "111",
          },
          "current": Immutable.Map {
            "helperText": "",
            "error": false,
            "pristine": true,
            "touched": false,
            "value": "111",
          },
        },
      },
    },
    "type": "addListItem",
  },
]
`;

exports[`remove 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": 1,
    "type": "removeListItem",
  },
]
`;

exports[`remove will trigger custom validation 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Object {
      "error": true,
      "helperText": "i am error",
    },
    "type": "validationResult",
  },
]
`;

exports[`remove will trigger validation 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Object {
      "error": false,
      "helperText": "",
    },
    "type": "validationResult",
  },
]
`;

exports[`returns props needed 1`] = `
Object {
  "props": Object {
    "add": [Function],
    "error": false,
    "helperText": "help me",
    "items": Array [
      Object {
        "first": Object {
          "error": "false",
          "helperText": "",
          "label": "First",
          "onBlur": [Function],
          "onChange": [Function],
          "value": "1",
        },
        "key": "fieldName.0",
        "second": Object {
          "error": "false",
          "helperText": "",
          "label": "Second",
          "onBlur": [Function],
          "onChange": [Function],
          "value": "2",
        },
      },
    ],
    "label": "the label",
    "remove": [Function],
    "updateIndex": [Function],
  },
  "setValue": [Function],
}
`;

exports[`updateIndex will dispatch action 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Object {
      "fromIndex": 0,
      "toIndex": 4,
    },
    "type": "updateListIndex",
  },
]
`;
